### ISSUE TITLE
Add Caller Context to the ObjectStore Trait for tracing and diagnostic

### ISSUE THREAD

**Issue by mingmwang** [apache/arrow-datafusion#3020]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

Some of the distributed storage systems like HDFS support API tracing.  The tracing info will log to audit log in server side and help uses to better diagnostic and understand the interaction between systems.

https://community.cloudera.com/t5/Community-Articles/Providing-tracing-with-Spark-Caller-Context/ta-p/245806

When we run Ballista/DataFusion on HDFS, it is better that we can provide the same capabilities. 
Basically, the Caller context will keep the information like who(app/job/stage/task) is calling the the ObjectStore and pass through those information from the compute engine to storage engine.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by alamb** [comment]
@crepererum  did something similar for tracing in  https://github.com/apache/arrow-datafusion/pull/2940 which might be valuable

Also here wrapping a `dyn Objectstore` with some other struct that implements `ObjectStore` could be interesting. https://github.com/apache/arrow-datafusion/issues/3019#issuecomment-1203932916

**Comment by crepererum** [comment]
I'm actually facing a similar issue with IOx. I think the `ObjecStore` methods should retrieve an additional parameter `ext: &Extensions` which can carry opaque extensions around (similar to #2940). DataFusion can than use this to pass the current context to the object store, which can then extract tracing information.

**Comment by mingmwang** [comment]
@crepererum I think maybe we just need to add a parameter of type String `` ext: Option<String>``.   Users of the ObjectStore can feel free to encode any tracing info to strings and pass through to the storage system.

**Comment by crepererum** [comment]
Some objects cannot be encoded via strings, so users then build some own registry objects (probably some semi-global hashmap), which is worse. That's the whole reason I've added the extension system in #2940, because I wanted to avoid that.
